src/Functions/FunctionDictGetKeys.cpp

```diff
@@ -0,0 +1,572 @@
+        /// of keys. Then rollback after use to free memory for next use.
+        Arena arena;
+        Block block;
+        while (executor.pull(block))
+        {
+            chassert(block.columns() >= num_keys + 1);
+
+            ColumnPtr attr_col = removeSpecialRepresentations(block.getByPosition(num_keys).column);
+            const size_t rows_in_block = attr_col->size();
+
+            std::vector<ColumnPtr> key_columns(num_keys);
+            for (size_t key_pos = 0; key_pos < num_keys; ++key_pos)
+            {
+                key_columns[key_pos] = removeSpecialRepresentations(block.getByPosition(key_pos).column);
+                chassert(key_columns[key_pos]->size() == rows_in_block);
+            }
+
+            for (size_t row_id = 0; row_id < rows_in_block; ++row_id)
+            {
+                const UInt128 value_hash = sipHash128AtRow(*attr_col, row_id);
+
+                /// Not in user given `values_column`
+                const auto * it = value_hash_to_bucket_id.find(value_hash);
+                if (it == value_hash_to_bucket_id.end())
+                    continue;
+
+                const size_t bucket_id = it->getMapped();
+
+                chassert(bucket_id < out.size());
+
+                /// In user given `values_column` but not needed
+                if (!is_missing[bucket_id])
+                    continue;
+
+                auto & mapped = out[bucket_id];
+                if (!mapped)
+                    mapped = std::make_shared<SerializedKeys>();
+
+                for (size_t key_pos = 0; key_pos < num_keys; ++key_pos)
+                {
+                    const auto & key_col = key_columns[key_pos];
+                    const char * begin = nullptr;
+                    StringRef ref = key_col->serializeValueIntoArena(row_id, arena, begin);
+
+                    const size_t old_size = mapped->size();
+                    const size_t need = old_size + ref.size;
+
+                    /// This is important. Otherwise, each repeated incremental `resize()` will cause
+                    /// repeated reallocations and copy which is very inefficient.
+                    if (need > mapped->capacity())
+                    {
+                        size_t cap = mapped->capacity();
+                        if (cap == 0)
+                            cap = 64;
+                        while (cap < need)
+                            cap *= 2;
+                        mapped->reserve(cap);
+                    }
+
+                    mapped->resize(need);
+                    std::memcpy(mapped->data() + old_size, ref.data, ref.size);
+
+                    const size_t alloc = static_cast<size_t>((ref.data - begin) + ref.size);
+
+                    /// This is important to rollback otherwise we will have double memory consumption.
+                    /// Additionally, just used memory is now hot in CPU cache which speeds up next serialization.
+                    arena.rollback(alloc);
```

Comment:
I wonder if there is an invariant to check that we rollback no more than what has been just allocated... 🤔